drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2346 @@
+
+	/* Prepare DAT before launching DAA. */
+	for (pos = 0; pos < data->maxdevs; pos++) {
+		if (olddevs & BIT(pos)) {
+			continue;
+		}
+
+		addr = i3c_addr_slots_next_free_find(&data->common.attached_dev.addr_slots,
+						     last_addr + 1);
+		if (addr == 0) {
+			return -ENOSPC;
+		}
+
+		p = odd_parity(addr);
+		last_addr = addr;
+		addr |= (p << 7);
+		sys_write32(DEV_ADDR_TABLE_DYNAMIC_ADDR(addr),
+			    config->regs + DEV_ADDR_TABLE_LOC(data->datstartaddr, pos));
+	}
+
+	pos = get_free_pos(data->free_pos);
+	if (pos < 0) {
+		LOG_ERR("%s: find free pos failed", dev->name);
+		return -ENOSPC;
+	}
+
+	ret = k_mutex_lock(&data->mt, K_MSEC(1000));
+	if (ret) {
+		LOG_ERR("%s: Mutex err (%d)", dev->name, ret);
+		return ret;
+	}
+	memset(xfer, 0, sizeof(struct dw_i3c_xfer));
+
+	xfer->ncmds = 1;
+	xfer->ret = -1;
+
+	cmd = &xfer->cmds[0];
+	cmd->cmd_hi = COMMAND_PORT_TRANSFER_ARG;
+	cmd->cmd_lo = COMMAND_PORT_TOC | COMMAND_PORT_ROC |
+		      COMMAND_PORT_DEV_COUNT(data->maxdevs - pos) | COMMAND_PORT_DEV_INDEX(pos) |
+		      COMMAND_PORT_CMD(I3C_CCC_ENTDAA) | COMMAND_PORT_ADDR_ASSGN_CMD;
+
+	start_xfer(dev);
+	ret = k_sem_take(&data->sem_xfer, K_MSEC(1000));
+	if (ret) {
+		LOG_ERR("%s: Semaphore err (%d)", dev->name, ret);
+		k_mutex_unlock(&data->mt);
+		return ret;
+	}
+
+	k_mutex_unlock(&data->mt);
+
+	if (data->maxdevs == cmd->rx_len) {
+		newdevs = 0;
+	} else {
+		newdevs = GENMASK(data->maxdevs - cmd->rx_len - 1, 0);
+	}
+	newdevs &= ~olddevs;
+
+	for (pos = find_lsb_set(newdevs); pos <= find_msb_set(newdevs); pos++) {
+		if (newdevs & BIT(pos)) {
+			add_slave_from_daa(dev, pos);
+		}
+	}
+
+	return 0;
+}
+
+static void dw_i3c_enable_controller(const struct dw_i3c_config *config, bool enable)
+{
+	uint32_t reg = sys_read32(config->regs + DEVICE_CTRL);
+
+	if (enable) {
+		reg |= DEV_CTRL_ENABLE;
+	} else {
+		reg &= ~DEV_CTRL_ENABLE;
+	}
+
+	sys_write32(reg, config->regs + DEVICE_CTRL);
+}
+
+/**
+ * @brief Get configuration of the I3C hardware.
+ *
+ * This provides a way to get the current configuration of the I3C hardware.
+ *
+ * This can return cached config or probed hardware parameters, but it has to
+ * be up to date with current configuration.
+ *
+ * @param[in] dev Pointer to controller device driver instance.
+ * @param[in] type Type of configuration parameters being passed
+ *                 in @p config.
+ * @param[in,out] config Pointer to the configuration parameters.
+ *
+ * Note that if @p type is @c I3C_CONFIG_CUSTOM, @p config must contain
+ * the ID of the parameter to be retrieved.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General Input/Output errors.
+ * @retval -ENOSYS If not implemented.
+ */
+static int dw_i3c_config_get(const struct device *dev, enum i3c_config_type type, void *config)
+{
+	const struct dw_i3c_config *dev_config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	int ret = 0;
+
+	if (config == NULL) {
+		ret = -EINVAL;
+		goto out_configure;
+	}
```

Comment:
redundant?

Reply:
I don't see where it's redundant

Reply:
Sounds like passing NULL would be a programming error, not something to be handled

Reply:
yeah.... I could change to a `__ASSERT_NO_MSG`

Reply:
i just removed it instead...